packages/grid/src/styles/vaadin-grid-base-styles.js

```diff
@@ -153,112 +151,178 @@ export const gridStyles = css`
+    border-top: 0;
+    box-shadow:
+      var(--_column-border-width) calc(var(--_row-border-width) * -1) 0 0
+        var(--vaadin-grid-border-color, var(--vaadin-border-color-secondary)),
+      var(--_column-border-width) 0 0 0 var(--vaadin-grid-cell-background-color, var(--vaadin-background-color));
   }
 
-  :host([theme~='column-borders']) {
-    --_column-border-width: var(--vaadin-grid-cell-border-width, 1px);
+  #footer:has(:is([frozen], [frozen-to-end])) [part~='first-footer-row-cell'] {
+    box-shadow:
+      var(--_column-border-width) calc(var(--_row-border-width) * -1) 0 0
+        var(--vaadin-grid-border-color, var(--vaadin-border-color-secondary)),
+      var(--_column-border-width) calc(var(--_row-border-width) * -1) 0 0
+        var(--vaadin-grid-cell-background-color, var(--vaadin-background-color));
   }
 
-  [part~='cell']:not([part~='last-column-cell'], [part~='details-cell']) {
-    border-inline-end: var(--_column-border-width, 0) solid
-      var(--vaadin-grid-cell-border-color, var(--vaadin-border-color-secondary));
+  [part~='body-cell']:not([part~='details-cell']) {
+    --_highlight-background-color: var(--vaadin-grid-row-highlight-background-color, transparent);
+    --_highlight-background-image: linear-gradient(
+      var(--_highlight-background-color),
+      var(--_highlight-background-color)
+    );
+    background:
+      var(--_hover-background-image, none), var(--_selected-background-image, none), var(--_highlight-background-image),
+      var(--vaadin-grid-cell-background-color, var(--vaadin-background-color));
   }
 
-  [part~='cell']:where(:not([part~='details-cell'], [part~='first-row-cell'])) {
-    border-top: var(--_row-border-width) solid
-      var(--vaadin-grid-cell-border-color, var(--vaadin-border-color-secondary));
+  [part~='cell']:focus-visible {
+    outline: 0;
   }
 
-  [part~='first-header-row-cell'] {
-    border-top: 0;
+  /* Keyboard navigation focus outline (row focus outline also used for drag'n'drop target indication) */
+  [part~='row']::after,
+  :host([navigating]) [part~='cell']::after {
+    position: absolute;
+    inset: calc(var(--_row-border-width) * -1) calc(var(--_column-border-width) * -1);
+    z-index: 3;
+    pointer-events: none;
+    outline: var(--vaadin-focus-ring-width) solid var(--vaadin-focus-ring-color);
+    outline-offset: calc(var(--vaadin-focus-ring-width) * -1);
   }
 
-  [part~='last-header-row-cell'] {
-    border-bottom: var(--_row-border-width, 1px) solid
-      var(--vaadin-grid-cell-border-color, var(--vaadin-border-color-secondary));
+  [part~='row']:focus-visible::after {
+    content: '';
   }
 
-  [part~='first-footer-row-cell'] {
-    border-top: var(--_row-border-width, 1px) solid
-      var(--vaadin-grid-cell-border-color, var(--vaadin-border-color-secondary));
+  [part~='row']:is(:focus, :focus-within) {
+    z-index: 3;
+  }
+
+  :host([navigating]) {
+    :is([part~='row'], [part~='cell']):is(:focus, :focus-within :focus-visible) {
```

Comment:
This combination of `:focus, :focus-within :focus-visible` is somewhat unclear, is it needed specifically for the focus button mode? Maybe could add a comment clarifying it.

Reply:
I’ll re-evaluate this once I get to redoing this in a separate PR. This is the result of me just trying to brute-force the visual tests to pass, without thinking clearly about the behavior or how the implementation works internally. 

I think there are some differences between Safari and other browsers regarding focus behavior. Other browsers seem to respect `:focus-visible` more reliable, whereas in Safari I need to use `:host([navigating]) :focus` instead.